app/assets/javascripts/discourse/app/modifiers/post-stream-viewport-tracker.js

```diff
@@ -0,0 +1,769 @@
+   * @param {HTMLElement} element - The wrapper element containing all posts
+   * @param {Object} options - Configuration options
+   * @param {Function} options.currentPostChanged - Callback when current post changes
+   * @param {Function} options.currentPostScrolled - Callback when scroll position within current post changes
+   * @param {number} options.headerOffset - Offset from top of viewport for site header
+   * @param {Object} options.screenTrack - Screen tracking service
+   * @param {Function} options.setCloakingBoundaries - Callback to update cloaking boundaries
+   * @returns {Function} Cleanup function that removes observers and event listeners
+   */
+  #setup = modifier(
+    (
+      element,
+      _,
+      {
+        currentPostChanged,
+        currentPostScrolled,
+        headerOffset,
+        screenTrack,
+        setCloakingBoundaries,
+      }
+    ) => {
+      this.#wrapperElement = element;
+
+      this.#currentPostChanged = currentPostChanged;
+      this.#currentPostScrolled = currentPostScrolled;
+      this.#headerOffset = headerOffset;
+      this.#screenTrackService = screenTrack;
+      this.#setCloakingBoundaries = setCloakingBoundaries;
+
+      this.#updateCloakOffset();
+      this.#setupEventListeners();
+
+      // intersection observers
+      this.#resetViewportObservers();
+      // eyeline
+      this.#setupEyelineDebugElement();
+
+      schedule("afterRender", () => {
+        // forces updates performed when the scroll is triggered
+        this.#scrollTriggered();
+      });
+
+      // clean-up
+      return () => {
+        this.#cloakingObserver?.disconnect();
+        this.#viewportObserver?.disconnect();
+
+        // clear event listeners
+        this.#setupEventListeners(false);
+        // remove the eyeline debug element
+        this.#setupEyelineDebugElement(false);
+      };
+    }
+  );
+
+  /**
+   * Cleans up resources when the tracker is destroyed
+   * Disconnects intersection observers and clears DOM references
+   */
+  destroy() {
+    // disconnect the intersection observers
+    this.#viewportObserver?.disconnect();
+    this.#cloakingObserver?.disconnect();
+
+    // clear DOM references
+    this.#observedPostElements.clear();
+  }
+
+  /**
+   * Returns a map of post numbers to their post models and DOM elements that are currently on screen
+   * @returns {Object<number, {post: Object, element: HTMLElement}>}
+   */
+  get postsOnScreen() {
+    return this.#postsOnScreen;
+  }
+
+  /**
+   * Returns the modifier for registering the bottom boundary element
+   * @returns {Modifier}
+   */
+  get registerBottomBoundary() {
+    return this.#registerBottomBoundary;
+  }
+
+  /**
+   * Returns the modifier for registering post elements
+   * @returns {Modifier}
+   */
+  get registerPost() {
+    return this.#registerPost;
+  }
+
+  /**
+   * Returns the main setup modifier
+   * @returns {Modifier}
+   */
+  get setup() {
+    return this.#setup;
+  }
```

Comment:
Wouldn't it be faster/clearer to just make `#setup` a public field?

Reply:
I'd like to keep it that way. I want these fields to be read only and keep things encapsulated